confluent_kafka/schema_registry/schema.py

```diff
@@ -1,116 +1,204 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Confluent Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
 import json
 
-from fastavro._schema import PRIMITIVES
-from fastavro.schema import schema_name, parse_schema
+PRIMITIVES = {'string', 'bytes', 'boolean', 'null',
+              'int', 'double', 'float', 'long'}
 
 
-def SchemaProvider(schema_str, schema_type='AVRO'):
+class Schema(object):
     """
-    Factory function for instantiating a particular type of schema.
+    An unregistered Schema.
 
     Args:
-        schema_str (str): Schema definition
+        schema (str): String representation of the schema.
+        references [Schema]: Schema's referenced by this one.
+        schema_type (str): The schema type: AVRO, PROTOBUF or JSON.
 
-    Keyword Args:
-        schema_type (str): Schema type
+    """
+    __slots__ = ['schema', 'references', 'schema_type']
 
-    Returns:
-         Schema: Schema
+    def __init__(self, schema, schema_type='AVRO', references=[]):
+        super(Schema, self).__init__()
 
-    """
-    if schema_type == 'AVRO':
-        return AvroSchema(schema_str)
-    raise TypeError("Unrecognized schema_type")
+        self.schema = schema
+        self.schema_type = schema_type
+        self.references = references
 
+    def __eq__(self, other):
+        return str(self) == str(other)
 
-class Schema(object):
-    """
-    Schema base class.
+    def __str__(self):
+        return json.dumps(self.to_json(), sort_keys=True)
 
-    Attributes:
-        schema_type (str): Declares the Schema type
+    # TODO: consider metaclass and/or decorators
+    # JSON duck type handling
+    def to_json(self):
+        """
+        Converts Schema to JSON serializable type (dict).
+        Use with the standard JSONEncoder with the JSONDuckEncoder.
 
-    Args:
-        schema_str (str): Schema string representation
-        registration_id (int, optional): Schema Registration ID. Defaults to -1(unregistered)
 
-    Keyword Arguments:
-        schema (str): Schema type to return. Defaults to ``AVRO``
+        Returns:
+            dict: JSON serializable Schema
 
-    Raises:
-        SchemaParseException if the schema can't be parsed.
+        """
+        schema_dict = {'schema': self.schema}
+        # new fields require schema registry 6.0+
+        if len(self.references) == 0 and self.schema_type == 'AVRO':
```

Comment:
this logic should be close schema registry, and is what i mean by closely coupling above (but again, of minor significance in this case, just good practice).